### PR TITLE
Support IMAP EXAMINE().

### PR DESCRIPTION
**Pull request by ameir**
This PR enables support for IMAP EXAMINE() for read-only folder selection.

### CODE REVIEW

File: src/Storage/Imap.php [hunk 1]
```diff
@@ -354,16 +354,31 @@ public function getFolders($rootFolder = null)
      * @throws Exception\RuntimeException
      * @throws \Zend\Mail\Protocol\Exception\RuntimeException
      */
-    public function selectFolder($globalName)
+    public function selectFolder($globalName, $examine = false)
```
Comment (Maks3w) [inline src/Storage/Imap.php]: Revert the changes of this method

File: src/Storage/Imap.php [hunk 2]
```diff
@@ -354,16 +354,31 @@ public function getFolders($rootFolder = null)
...
-        if (!$this->protocol->select($this->currentFolder)) {
+        $method = ($examine ? 'examine' : 'select');
+        if (!$this->protocol->$method($this->currentFolder)) {
             $this->currentFolder = '';
             throw new Exception\RuntimeException('cannot change folder, maybe it does not exist');
         }
     }
 
     /**
+     * examine given folder
```
Comment (Maks3w) [inline src/Storage/Imap.php]: Add a more detailed description about why this method is different from selectFolder

**Comment by Maks3w** [comment]
Need tests

File: src/Storage/Imap.php (hunk 1 above)
Comment (ameir) [inline src/Storage/Imap.php]: Would you rather have `examineFolder()` be identical to `selectFolder()` aside from the examine/select?  i.e. copy/pasting `selectFolder()` and using examine instead?

File: src/Storage/Imap.php (hunk 2 above)
Comment (ameir) [inline src/Storage/Imap.php]: There's nothing better than quoting the RFC:

```
The EXAMINE command is identical to SELECT and returns the same
      output; however, the selected mailbox is identified as read-only.
      No changes to the permanent state of the mailbox, including
      per-user state, are permitted; in particular, EXAMINE MUST NOT
      cause messages to lose the \Recent flag.
```

Anyone who uses it should know what it is, just as they know when to use `SELECT`.

**Comment by weierophinney** [comment]
Needs a rebase, @ameir — as well as tests.

**Comment by ameir** [comment]
I added some basic tests to this.  Let me know how it looks.

File: test/Storage/ImapTest.php [hunk 3]
```diff
@@ -594,6 +594,13 @@ public function testSelect()
         $this->assertEquals($status['exists'], 7);
     }
 
+    public function testStorageSelect()
+    {
+        $mail = new Storage\Imap($this->params);
```
Comment (Maks3w) [inline test/Storage/ImapTest.php]: You could inject a protocol and spy the expected method is invoked

Comment (ameir) [inline test/Storage/ImapTest.php]: Do you have an example of that somewhere for reference?

Comment (Maks3w) [inline test/Storage/ImapTest.php]: https://phpunit.de/manual/current/en/test-doubles.html

https://github.com/zendframework/zend-mail/blob/51f5f729c541adc18dfa6497a45be32d1a3d63ae/test/Header/HeaderWrapTest.php#L24-L27

File: src/Storage/Imap.php [hunk 4]
```diff
@@ -372,16 +372,31 @@ public function getFolders($rootFolder = null)
      * @throws Exception\RuntimeException
      * @throws Protocol\Exception\RuntimeException
      */
-    public function selectFolder($globalName)
+    public function selectFolder($globalName, $readOnly = false)
```
Comment (weierophinney) [inline src/Storage/Imap.php]: Potential BC break for any extensions to this class, but only if those extensions are typehinting the additional, optional argument. Will need to draw attention to this in the release notes.

File: src/Storage/Imap.php (hunk 2 above)
Comment (weierophinney) [inline src/Storage/Imap.php]: Please add a `@see <link to RFC section>` to the docblock, and/or quote the RFC to make this clear.

File: test/Storage/ImapTest.php (hunk 3 above)
Comment (weierophinney) [inline test/Storage/ImapTest.php]: Please incorporate the feedback from @Maks3w

**Comment by weierophinney** [comment]
Closing; author has not incorporated requests for changes.

If you would like to re-open, please incorporate the feedback and re-submit.

Thanks!
